### PR TITLE
BUG: MultiIndex.difference incorrectly raising TypeError when indexes contain non-sortable entries

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/multi/test_setops.py [hunk 1]
```diff
@@ -216,11 +216,10 @@ def test_difference_sort_incomparable():
...
     other = MultiIndex.from_product([[3, pd.Timestamp("2000"), 4], ["c", "d"]])
     # sort=None, the default
-    # MultiIndex.difference deviates here from other difference
-    # implementations in not catching the TypeError
-    msg = "'<' not supported between instances of 'Timestamp' and 'int'"
-    with pytest.raises(TypeError, match=msg):
+    msg = "sort order is undefined for incomparable objects"
+    with tm.assert_produces_warning(RuntimeWarning, match=msg):
         result = idx.difference(other)
+    tm.assert_index_equal(result, idx)
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/multi/test_setops.py:222]: behavior change? needs a release note?

**Comment by jorisvandenbossche** [comment]
Can you check the performance impact here? (xref https://github.com/pandas-dev/pandas/pull/41929#issuecomment-860514675)

**Comment by jbrockmendel** [comment]
> Can you check the performance impact here?

ran asvs overnight and no obvious standouts.  re-running a subset now to be on the safe side

**Comment by jbrockmendel** [comment]
yep, all noise

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**